### ISSUE TITLE
Conda cannot be activated on Windows if AppLocker is used

### ISSUE THREAD

**Issue by pnsaevik** [conda/conda#13610]
### Checklist

- [X] I added a descriptive title
- [X] I searched open reports and couldn't find a duplicate

### What happened?

The problem is described in the issue https://github.com/ContinuumIO/anaconda-issues/issues/9450

In short, when AppLocker is used on a Windows system (which is a common company-wide policy), conda is blocked.

The user gets the following unhelpful response: "This program is blocked by group policy. For more information, contact your system administrator."

The problem is that conda activation creates a temporary script with a random name, puts it in the 'Temp' folder, and tries to execute it. This looks too much like malware to AppLocker, so it gets blocked.

Something similar happens if one tries to activate conda through PowerShell.

I don't know what would be the best solution to this problem. A workaround is outlined in https://github.com/ContinuumIO/anaconda-issues/issues/9450.

### Additional Context

![image](https://github.com/conda/conda/assets/12728107/f79f3edf-a654-4275-a360-2fe30b93bb71)

**Comment by travishathaway** [comment]
Hi @pnsaevik,

Thanks for filing this detailed issue, and we also really appreciate the proposed work around that you have shared with others. It appears that this problem is widespread enough to warrant potentially updating conda itself to use a solution like yours.

At this point, I think it would make sense to work together to draft a pull request in order to further discuss your solution and see how we could possibly integrate this into conda. Would you be able/willing to create this pull request yourself? Is there anything we could help with?

I'm going to share this issue with other maintainers so they can chime in on the discussion too.

**Comment by skupr-anaconda** [comment]
Could you please clarify if it will behave in the same way as before if you run the `Anaconda Prompt` as **Administrator**?
![Screenshot 2024-02-21 at 10 39 37](https://github.com/conda/conda/assets/79282962/81967def-2571-46cb-abfb-ab0c505d7ddf)

**Comment by pnsaevik** [comment]
Hi @travishathaway, I'm quite busy the coming month, but after that I might be able to contribute with some code/pull request. But there are other challenges as well. My proposed workaround is hackish and might be patched by Windows updates. While local variants of the workaround are possible, a standardized, open-source implementation might be vulnerable to exploits. I fear that a "proper" solution would require a more substantial rewrite of the activation system, at the risk of breaking established workflows for other users. If people in the development team are familiar with such issues, we should discuss alternative solutions before we start writing code.

**Comment by pnsaevik** [comment]
Hi @skupr-anaconda, the issue disappears when running Anaconda Prompt as administrator.

**Comment by travishathaway** [comment]
@pnsaevik,

I totally agree with you. I'm not entirely sure about the development team's experience with App Locker, but we are constantly striving to improve our Windows compatibility issues considering how many of our users use this operating system.

Let's discuss here first before writing any code.

**Comment by pnsaevik** [comment]
Some questions to start the discussion:

1. What are the purposes of activation, i.e., what should the activation procedure be able to do?
2. Why is a temporary .bat or .ps1 file required in the first place? Could the same effect be obtained by feeding a config file to a static script?
3. Does the temporary script have to be random scripts in the temp folder? It would be easier for administrators to relax AppLocker rules if the generated script files were either signed, had a special naming convention, or were put in a special folder.
4. If conda generated scripts with a special naming convention, and local administrators allowed the execution of these scripts, what would be the security implications? Could an attacker, knowing that the machine allowed certainly named scripts, put a malicious script with the same naming pattern on the computer, and execute it?

**Comment by kenodegard** [comment]
> 1. What are the purposes of activation, i.e., what should the activation procedure be able to do?

Activation boils down to modifying your `%PATH%` (and setting some additional environment variables) to contain library paths from the activated environment.

> 2. Why is a temporary .bat or .ps1 file required in the first place?

Ideally we would do what is done in POSIX where we are able to evaluate the stdout of a command (for POSIX shells we run `eval "$(conda shell.posix activate $@)"`). Since CMD doesn't support this temporary scripts are used instead. I'm not fully versed in PowerShell so I'm not sure there.

> Could the same effect be obtained by feeding a config file to a static script?

Possibly, the activation process and scripts haven't been tinkered with in a while so it's likely time to revisit the entire thing.

> 3. Does the temporary script have to be random scripts in the temp folder?

The randomness is to allow parallel runs of `conda activate`, see https://github.com/conda/conda/pull/9477.

> It would be easier for administrators to relax AppLocker rules if the generated script files were either signed, had a special naming convention, or were put in a special folder.

The script as defined today will always be `%TMP%/conda-%RANDOM%/conda.tmp`.

> 4. If conda generated scripts with a special naming convention, and local administrators allowed the execution of these scripts, what would be the security implications? Could an attacker, knowing that the machine allowed certainly named scripts, put a malicious script with the same naming pattern on the computer, and execute it?

Yes, it is possible for an attacker to inject malicious scripts into the activation process if they have write access to the conda environment. This can be done via activation/deactivation scripts, see https://docs.conda.io/projects/conda/en/latest/dev-guide/deep-dives/activation.html#activation-deactivation-scripts.

On the other hand, I don't see how an attacker could benefit from placing a malicious script in `%TMP%`. If an attacker was able to predict and create the randomized temporary script ahead of time and had write access conda would first attempt to overwrite that script. If it failed to overwrite said script we move on and try another randomized temporary script: https://github.com/conda/conda/blob/2ca3d3532a176b4ece917d4cf39ab5ca47ffec6c/conda/shell/condabin/_conda_activate.bat#L20-L28

**Comment by pnsaevik** [comment]
Thanks @kenodegard, this is useful background information. Some comments:

1. If the script only sets environment variables, this should be possible to achieve without .bat files.
2. Take a look at https://stackoverflow.com/questions/48467703/reading-key-value-pair-using-batch-files. Is it possible that conda, instead of calling a recently generated .bat file, could set environment variables using the mechanism proposed in the StackOverflow question?
3. Your response to Q3 is only partially correct. The file `%TMP%/conda-%RANDOM%/conda.tmp` only contains the *name* of the temporary .bat file, which is named something like `%TMP%/%RANDOM%.bat`. So it's hard to generate an AppLocker rule based on this.
4. I'm no security expert, but I can imagine the following scenario: Through a security vulnerability, an attacker is able to download an executable file into the temp folder and run it. Usually AppLocker would block this attack, but if `TEMP` is added as an AppLocker exception the attack can proceed.

**Comment by kenodegard** [comment]
Not stale
